plugins/module_utils/user.py

```diff
@@ -107,37 +111,56 @@ def get_existing_authentication(cursor, user, host):
     if 'mariadb' in srv_type[0].lower():
         # before MariaDB 10.2.19 and 10.3.11, "password" and "authentication_string" can differ
         # when using mysql_native_password
-        cursor.execute("""select plugin, auth from (
-            select plugin, password as auth from mysql.user where user=%(user)s
-            and host=%(host)s
-            union select plugin, authentication_string as auth from mysql.user where user=%(user)s
-            and host=%(host)s) x group by plugin, auth limit 2
-        """, {'user': user, 'host': host})
+        if host:
+            cursor.execute("""select plugin, auth from (
+                select plugin, password as auth from mysql.user where user=%(user)s
+                and host=%(host)s
+                union select plugin, authentication_string as auth from mysql.user where user=%(user)s
+                and host=%(host)s) x group by plugin, auth
+            """, {'user': user, 'host': host})
+        else:
+            cursor.execute("""select plugin, auth from (
+                select plugin, password as auth from mysql.user where user=%(user)s
+                union select plugin, authentication_string as auth from mysql.user where user=%(user)s
+                ) x group by plugin, auth
+            """, {'user': user})
     else:
-        cursor.execute("""select plugin, authentication_string as auth
-            from mysql.user where user=%(user)s and host=%(host)s
-            group by plugin, authentication_string limit 2""", {'user': user, 'host': host})
+        if host:
+            cursor.execute("""select plugin, authentication_string as auth
+                from mysql.user where user=%(user)s and host=%(host)s
+                group by plugin, authentication_string""", {'user': user, 'host': host})
+        else:
+            cursor.execute("""select plugin, authentication_string as auth
+                from mysql.user where user=%(user)s
+                group by plugin, authentication_string""", {'user': user})
+
     rows = cursor.fetchall()
 
+    if len(rows) == 0:
+        return []
+
     # Mysql_info use a DictCursor so we must convert back to a list
     # otherwise we get KeyError 0
     if isinstance(rows, dict):
         rows = list(rows.values())
```

Comment:
The code above should be executed per row, not on all rows. With a DictCursor, I expect `fetchall()` to return a list of dicts, for example `[{"plugin": "foo", "auth": "bar"}]`.
When fixing this here, there is no need to check for tuple/dict in the loop later.

Reply:
@betanummeric I'm force to check for tuple and dict because this module utils is used from different module with different cursors :(

Reply:
`rows` is either type `list[dict]` or type `list[tuple]`, depending on the cursor. But `isinstance(rows, dict)` will never be true, therefore this code is broken.
Apparently the intention was to convert `list[dict]` to `list[list]`, which is compatible with `list[tuple]`. In other words, convert a dict-cursor result to a tuple-cursor result.
If we make that work again, we guarantee tuple-style rows and don't need to skip for dict or tuple.

Reply:
I don't write Python frequently enough, I keep writing silly stuff. Your feedback is really appreciated @betanummeric. I've fixed that with a list comprehension, I'm running tests and will add it to the PR when everything is green.